pkg/controllers/machine/lifecycle/launch.go

```diff
@@ -0,0 +1,147 @@
+	PopulateMachineDetails(machine, created)
+	machine.StatusConditions().MarkTrue(v1alpha5.MachineLaunched)
+	return reconcile.Result{}, nil
+}
+
+func (l *Launch) linkMachine(ctx context.Context, machine *v1alpha5.Machine) (*v1alpha5.Machine, error) {
+	created, err := l.cloudProvider.Get(ctx, machine.Annotations[v1alpha5.MachineLinkedAnnotationKey])
+	if err != nil {
+		if cloudprovider.IsMachineNotFoundError(err) {
+			if err = l.kubeClient.Delete(ctx, machine); err != nil {
+				return nil, client.IgnoreNotFound(err)
+			}
+			logging.FromContext(ctx).Debugf("garbage collected machine with no cloudprovider representation")
+			metrics.MachinesTerminatedCounter.With(prometheus.Labels{
+				metrics.ReasonLabel:      "garbage_collected",
+				metrics.ProvisionerLabel: machine.Labels[v1alpha5.ProvisionerNameLabelKey],
+			}).Inc()
+			return nil, nil
+		}
+		machine.StatusConditions().MarkFalse(v1alpha5.MachineLaunched, "LinkFailed", truncateMessage(err.Error()))
+		return nil, fmt.Errorf("linking machine, %w", err)
+	}
+	logging.FromContext(ctx).Debugf("linked machine")
+	return created, nil
+}
+
+func (l *Launch) launchMachine(ctx context.Context, machine *v1alpha5.Machine) (*v1alpha5.Machine, error) {
+	instanceTypeRequirement, _ := lo.Find(machine.Spec.Requirements, func(req v1.NodeSelectorRequirement) bool { return req.Key == v1.LabelInstanceTypeStable })
+	logging.FromContext(ctx).With("pods", machine.Spec.Resources.Requests[v1.ResourcePods],
+		"resources", machine.Spec.Resources.Requests, "instance-types", instanceTypeList(instanceTypeRequirement.Values)).Infof("launching machine")
+	created, err := l.cloudProvider.Create(ctx, machine)
+	if err != nil {
+		if cloudprovider.IsInsufficientCapacityError(err) {
+			logging.FromContext(ctx).Error(err)
+			if err = l.kubeClient.Delete(ctx, machine); err != nil {
+				return nil, client.IgnoreNotFound(err)
+			}
+			metrics.MachinesTerminatedCounter.With(prometheus.Labels{
+				metrics.ReasonLabel:      "insufficient_capacity",
+				metrics.ProvisionerLabel: machine.Labels[v1alpha5.ProvisionerNameLabelKey],
+			}).Inc()
+			return nil, nil
+		}
+		machine.StatusConditions().MarkFalse(v1alpha5.MachineLaunched, "LaunchFailed", truncateMessage(err.Error()))
+		return nil, fmt.Errorf("creating machine, %w", err)
+	}
+	logging.FromContext(ctx).Debugf("launched machine")
+	return created, nil
+}
+
+func PopulateMachineDetails(machine, retrieved *v1alpha5.Machine) {
+	machine.Labels = lo.Assign(
+		scheduling.NewNodeSelectorRequirements(machine.Spec.Requirements...).Labels(),
+		machine.Labels,
+		retrieved.Labels,
+		map[string]string{
+			v1alpha5.MachineNameLabelKey: machine.Name,
+		},
+	)
+	machine.Annotations = lo.Assign(machine.Annotations, retrieved.Annotations)
+	machine.Status.ProviderID = retrieved.Status.ProviderID
+	machine.Status.Allocatable = retrieved.Status.Allocatable
+	machine.Status.Capacity = retrieved.Status.Capacity
```

Comment:
Any reason we can't do machine.Status = retrieved.Status?

Reply:
I don't think we want to override the conditions here.